File: Java/21_Binary Search Tree-I/M1008. Construct Binary Search Tree from Preorder Traversal.java
// 1008. Construct Binary Search Tree from Preorder Traversal
// https://leetcode.com/problems/construct-binary-search-tree-from-preorder-traversal/

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) {
        this.val = val;
    }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {

    public TreeNode bstFromPreorder(int[] preorder) {
        int[] index = { 0 };
        return construct(preorder, index, Integer.MAX_VALUE);
    }

    private TreeNode construct(int[] preorder, int[] index, int bound) {
        if (
            index[0] == preorder.length || preorder[index[0]] > bound
        ) return null;
        TreeNode root = new TreeNode(preorder[index[0]++]);
        root.left = construct(preorder, index, root.val);
        root.right = construct(preorder, index, bound);
        return root;
    }

    public void printTree(TreeNode root) {
        public List<Integer> toLevelOrder(TreeNode root) {
                List<Integer> result = new ArrayList<>();
                Queue<TreeNode> q = new LinkedList<>();
                q.offer(root);
                while (!q.isEmpty()) {
                    TreeNode node = q.poll();
                    if (node == null) {
                        result.add(null);
                        continue;
                    }
                    result.add(node.val);
                    if (node.left != null || node.right != null) {
                        q.offer(node.left);
                        q.offer(node.right);
                    }
                }
                return result;
            }
}

class Main {

    public static void main(String[] args) {
        TreeNode root = new Solution().bstFromPreorder(new int[] { 1, 3 });
        System.out.println(sol.toLevelOrder(root)); // Output: [1,null,3]
    }
}

File: JavaScript/21_Binary Search Tree-I/M1008. Construct Binary Search Tree from Preorder Traversal.js
/**
 * 1008. Construct Binary Search Tree from Preorder Traversal
 * https://leetcode.com/problems/construct-binary-search-tree-from-preorder-traversal/
 *
 * Type             | Details
 * -----------------+----------------------------------------------------
 * Time Complexity  | O(N) – Each element is visited once
 * Space Complexity | O(H) – Recursion stack
 */

/**
 * Definition for a binary tree node.
 */
/**
 * Creates a tree node
 */
function createTreeNode(val = 0, left = null, right = null) {
    return { val, left, right };
}

/**
 * Construct BST from preorder traversal
 * @param {number[]} preorder - Preorder traversal array
 * @returns {TreeNode} - Root of the constructed BST
 */
function bstFromPreorder(preorder) {
    const index = { val: 0 };
    return construct(preorder, index, Infinity);
}

function construct(preorder, index, bound) {
    if (index.val === preorder.length || preorder[index.val] > bound) {
        return null;
    }
    const root = createTreeNode(preorder[index.val++]);
    root.left = construct(preorder, index, root.val);
    root.right = construct(preorder, index, bound);
    return root;
}

/**
 * Convert tree to level order array
 * @param {TreeNode} root
 * @returns {(number|null)[]}
 */
function toLevelOrder(root) {
    const result = [];
    const queue = [root];
    while (queue.length > 0) {
        const node = queue.shift();
        if (node === null) {
            result.push(null);
            continue;
        }
        result.push(node.val);
        if (node.left !== null || node.right !== null) {
            queue.push(node.left);
            queue.push(node.right);
        }
    }
    return result;
}

// Test cases
const root = bstFromPreorder([8, 5, 1, 7, 10, 12]);
console.log(toLevelOrder(root)); // Output: [8, 5, 10, 1, 7, null, 12]

const root2 = bstFromPreorder([1, 3]);
console.log(toLevelOrder(root2)); // Output: [1, null, 3]